### PR TITLE
Add support for flow typecast comments

### PR DESCRIPTION
**Pull request by swac**
This attempts to fix the bug described in #719. Previously, Prettier was removing parentheses in expressions like: 

```
new (window.Request /*: Class<Request> */)("")
```

turning them into:

```
new window.Request /*: Class<Request> */("")
```

which no longer satisfies Flow's syntax. With this change, we detect these typecast expressions and attempt to keep the parentheses intact in such cases.

This is my first time contributing to Prettier, so I'm probably making some egregious mistakes!

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: src/language-js/flow-comments.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+  return (
+    node.extra &&
+    node.extra.parenthesized &&
+    !!node.trailingComments &&
+    node.trailingComments[0].value.startsWith(":")
+  );
+}
+
+module.exports = hasFlowShorthandAnnotationComment;
```
Comment (j-f1) [inline src/language-js/flow-comments.js]: Can you rename this file to `utils.js` and export this as an object (i.e. `{ hasFlowShorthandAnnotationComment }`)?

File: src/language-js/flow-comments.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+function hasFlowShorthandAnnotationComment(node) {
+  // https://flow.org/en/docs/types/comments/
+  // Syntax example: const r = new (window.Request /*: Class<Request> */)("");
+
+  return (
+    node.extra &&
+    node.extra.parenthesized &&
+    !!node.trailingComments &&
+    node.trailingComments[0].value.startsWith(":")
```
Comment (j-f1) [inline src/language-js/flow-comments.js]: Can the comment be formatted as `/* : foo */`? If so, perhaps use `.value.trimLeft()`?

File: src/language-js/flow-comments.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+"use strict";
+
+function hasFlowShorthandAnnotationComment(node) {
+  // https://flow.org/en/docs/types/comments/
+  // Syntax example: const r = new (window.Request /*: Class<Request> */)("");
+
+  return (
+    node.extra &&
+    node.extra.parenthesized &&
+    !!node.trailingComments &&
```
Comment (j-f1) [inline src/language-js/flow-comments.js]: Can this just be `node.trailingComments`?

File: src/language-js/printer-estree.js [hunk 4]
```diff
@@ -2802,14 +2812,27 @@ function printPathNoParens(path, options, print, args) {
...
+      // Flow supports a comment syntax for specifying type annotations: https://flow.org/en/docs/types/comments/.
+      // Unfortunately, its parser doesn't differentiate between comment annotations and regular
+      // annotations when producing an AST. So to preserve parentheses around type casts that use
+      // the comment syntax, we need to hackily read the source itself to see if the code contains
+      // a type annotation comment.
+      const commentSyntax =
+        value &&
+        value.typeAnnotation &&
+        value.typeAnnotation.range &&
+        options.originalText.substr(value.typeAnnotation.range[0], 3) === "/*:";
```
Comment (j-f1) [inline src/language-js/printer-estree.js]: Similar question to the one in `flow-comments.js` — should this be checking against the regex `/^/\*\s*:/`?

File: src/language-js/flow-comments.js (hunk 2 above)
Comment (swac) [reply src/language-js/flow-comments.js]: Good point, it seems whitespace between `/*` and `:` is indeed allowed.

**Comment by lydell** [comment]
We need to be careful with that whitespace trimming though, because newlines do not seem to be supported. A goal of Prettier is to never break your code. Here's a case where it could break: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVxhgIIGMAuArgIYxj4CeADgKZjEB2Dc+x+AlnAwFyoBuxAE70wobmAaEAtgCMawlGAC8YAKzpMYAHIsRlWvSYs2nHvyFgZokKnEB+SbPnIsK9ZoAqgimEE0pcHzsDADm5AAWdAw0CDDBdHJQcH4RdLhw8AxggMgEYBRwhADkMGQhNPh61HTygskAhOgCwrjKIgA8lkA

File: src/language-js/utils.js [hunk 5]
```diff
@@ -8,7 +8,15 @@ function hasFlowShorthandAnnotationComment(node) {
...
-    node.trailingComments[0].value.trimLeft().startsWith(":")
+    // We intentionally only match spaces and tabs here instead of any whitespace because
+    // Flow annotation comments cannot be split across lines. For example:
+    //
+    // (this /*
+    // : any */).foo = 5;
+    //
+    // is not picked up by Flow, so removing the newline would create a type annotation
+    // that the user did not intend to create.
+    node.trailingComments[0].value.match(/^[ \t]*:/)
```
Comment (lydell) [inline src/language-js/utils.js]: According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) `\s` is equivalent to:

```
 \f\n\r\t\v\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff
```

I just tried it, and Flow allows any of those except the JS line terminators `\r`, `\n`, `\u2028` and `\u2029`.

This regex might do the trick: `/^(?:(?=.)\s)*:/`

File: src/language-js/utils.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+function hasFlowShorthandAnnotationComment(node) {
+  // https://flow.org/en/docs/types/comments/
+  // Syntax example: const r = new (window.Request /*: Class<Request> */)("");
+
+  return (
+    node.extra &&
+    node.extra.parenthesized &&
+    node.trailingComments &&
+    // We intentionally only match spaces and tabs here instead of any whitespace because
```
Comment (lydell) [inline src/language-js/utils.js]: `We match any whitespace except line terminators because`

File: src/language-js/printer-estree.js [hunk 7]
```diff
@@ -151,7 +152,16 @@ function genericPrint(path, options, printPath, args) {
...
-    parts.push(")");
+    const node = path.getValue();
+    if (hasFlowShorthandAnnotationComment(node)) {
+      parts.push(" /*");
+      parts.push(node.trailingComments[0].value.trimLeft());
+      parts.push("*/");
+      node.trailingComments[0].printed = true;
+      parts.push(")");
+    } else {
+      parts.push(")");
```
Comment (j-f1) [inline src/language-js/printer-estree.js]: Can this line be pulled out of both blocks and placed after the `if` statement?

File: src/language-js/printer-estree.js [hunk 8]
```diff
@@ -2802,14 +2812,34 @@ function printPathNoParens(path, options, print, args) {
...
+          .substring(value.typeAnnotation.range[0])
+          .match(/^\/\*\s*:/);
       return concat([
         "(",
         path.call(print, "expression"),
-        ": ",
+        commentSyntax ? " /*: " : ": ",
         path.call(print, "typeAnnotation"),
-        ")"
+        commentSyntax ? " */)" : ")"
```
Comment (j-f1) [inline src/language-js/printer-estree.js]: Similar thing here: how about splitting these two lines into an optional comment character and then the required `:` and `)`?

File: src/language-js/utils.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    node.extra.parenthesized &&
+    node.trailingComments &&
+    // We match any whitespace except line terminators because
+    // Flow annotation comments cannot be split across lines. For example:
+    //
+    // (this /*
+    // : any */).foo = 5;
+    //
+    // is not picked up by Flow, so removing the newline would create a type annotation
+    // that the user did not intend to create.
```
Comment (j-f1) [inline src/language-js/utils.js]: Do you want to reference https://github.com/facebook/flow/issues/7050 here?

**Comment by j-f1** [comment]
Looks like you’ve just got one lint error left.

**Comment by swac** [comment]
Thanks so much for reviewing! It looks like the commits in master are squashed, so I don't need to squash my commits in this branch, correct?

**Comment by j-f1** [comment]
That’s correct! I’ll merge tomorrow if nobody has any objections.

**Pull request merged**